indexing-service/src/main/java/org/apache/druid/indexing/overlord/hrtr/HttpRemoteTaskRunner.java

```diff
@@ -1021,92 +1043,164 @@ public void unregisterListener(String listenerId)
+              taskItem = ti;
+              break;
             }
-            catch (Throwable th) {
-              log.makeAlert(th, "Exception while trying to assign task")
-                 .addData("taskId", taskId)
-                 .emit();
 
-              if (taskItem != null) {
-                taskComplete(taskItem, null, TaskStatus.failure(taskId));
-              }
+            immutableWorker = findWorkerToRunTask(ti.getTask());
+            if (immutableWorker == null) {
+              continue;
+            }
+
+            String prevUnackedTaskId = workersWithUnacknowledgedTask.putIfAbsent(
+                immutableWorker.getWorker().getHost(),
+                taskId
+            );
+            if (prevUnackedTaskId != null) {
+              log.makeAlert(
+                  "Found worker[%s] with unacked task[%s] but still was identified to run task[%s].",
+                  immutableWorker.getWorker().getHost(),
+                  prevUnackedTaskId,
+                  taskId
+              ).emit();
+            }
+
+            ti.setState(HttpRemoteTaskRunnerWorkItem.State.PENDING_WORKER_ASSIGN);
+            taskItem = ti;
+            break;
+          }
+
+          if (taskItem == null) {
+            // Either no pending task is found or no suitable worker is found for any of the pending tasks.
+            // statusLock.notifyAll() is called whenever a new task shows up or if there is a possibility for a task
+            // to successfully get worker to run, for example when a new worker shows up, a task slot opens up
+            // because some task completed etc.
+            statusLock.wait(TimeUnit.MINUTES.toMillis(1));
+            continue;
+          }
+        }
+
+        String taskId = taskItem.getTaskId();
+
+        if (taskItem.getTask() == null) {
+          log.makeAlert("No Task obj found in TaskItem for taskID[%s]. Failed.", taskId).emit();
+          taskComplete(taskItem, null, TaskStatus.failure(taskId));
+          continue;
+        }
+
+        try {
+          if (immutableWorker == null) {
+            throw new ISE("NULL immutableWorker");
+          }
 
-              return;
+          // this will send HTTP request to worker for assigning task
+          if (!runTaskOnWorker(taskItem, immutableWorker.getWorker().getHost())) {
+            if (taskItem.getState() == HttpRemoteTaskRunnerWorkItem.State.PENDING_WORKER_ASSIGN) {
+              taskItem.revertStateFromPendingWorkerAssignToPending();
             }
           }
         }
-    );
+        catch (InterruptedException ex) {
+          log.info("Got InterruptedException while assigning task[%s].", taskId);
+          throw ex;
+        }
+        catch (Throwable th) {
+          log.makeAlert(th, "Exception while trying to assign task")
+             .addData("taskId", taskId)
+             .emit();
+
+          taskComplete(taskItem, null, TaskStatus.failure(taskId));
```

Comment:
This is called outside of `statusLock`.

Reply:
yes taskCompleted(..) has to be called outside of statusLock .. I have added comments please take a look.

Reply:
Oh, sorry. I misread codes that `taskComplete` requires to be called under `statusLock`. Thanks for adding comments.